### PR TITLE
Make WindowsUtil available for use

### PR DESCRIPTION
**Pull request by res0nance**
This change is came about as a follow up to jenkinsci/ec2-plugin#352 where this class had to be copied to ensure proper character escaping on Windows and would likely be useful elsewhere.

### Proposed changelog entries

* Internal: Move WindowsUtil out of test

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

File: core/src/main/java/hudson/os/WindowsUtil.java [hunk 1]
```diff
@@ -101,10 +99,12 @@
      * @return the newly created junction point
      * @throws IOException if the call to mklink exits with a non-zero status code
      * @throws InterruptedException if the call to mklink is interrupted before completing
-     * @throws AssertionError if this method is called on a non-Windows platform
+     * @throws UnsupportedOperationException if this method is called on a non-Windows platform
      */
     public static @Nonnull File createJunction(@Nonnull File junction, @Nonnull File target) throws IOException, InterruptedException {
-        assertTrue(Functions.isWindows());
+        if(Functions.isWindows() == false) {
```
Comment (jvz) [inline core/src/main/java/hudson/os/WindowsUtil.java]:
```suggestion
        if (!Functions.isWindows()) {
```

**Review by jvz — APPROVED** [review]
Looks good to me. I leave an open question on whether or not all the methods in this class should be unrestricted as is.

**Comment by varyvol** [comment]
@res0nance What do you think about checking for a Windows platform in all methods? I know it was not done before but it might be a good moment.

**Comment by res0nance** [comment]
@varyvol If we are using this util to quote arguments for running on an agent and we checked for windows platforms with linux master it will throw

**Comment by varyvol** [comment]
@res0nance as I see it, if that method is being used for situations other than Windows....then it shouldn't be in a class called WindowsUtil.

**Comment by res0nance** [comment]
The current usage was to use it to quote arguments for the EC2 plugin's winrm launcher

**Comment by jvz** [comment]
The only Windows-specific stuff in this class are the functions to support directory junctions, and that's only used in a unit test so far. This tool could be useful when your master is on Linux but you have a Windows agent.

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
The change itself looks good to me, but it lacks Javadoc. Also., it would be better to move the new code somewhere under https://github.com/jenkinsci/jenkins/tree/master/core/src/main/java/jenkins/util instead of the old `hudson` package

File: core/src/main/java/jenkins/util/WindowsUtil.java [hunk 2]
```diff
@@ -34,15 +34,19 @@
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
-import static org.junit.Assert.assertTrue;
-
-// adapted from:
-// https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/
+/**
+ * Utilities for the Windows Platform
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/util/WindowsUtil.java]:
```suggestion
 * Utilities for the Windows Platform.
```

File: core/src/main/java/jenkins/util/WindowsUtil.java [hunk 3]
```diff
@@ -34,15 +34,19 @@
...
+ * adapted from:
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/util/WindowsUtil.java]:
```suggestion
 * Adapted from:
```

File: core/src/main/java/jenkins/util/WindowsUtil.java [hunk 4]
```diff
@@ -34,15 +34,19 @@
...
+ * https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/
+ */
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/util/WindowsUtil.java]:
```suggestion
 *
 * @since TODO
 */
```

**Review by oleg-nenashev — APPROVED** [review]
Just minor suggestions in Javadoc,, LGTM

**Review by oleg-nenashev — APPROVED** [review]
Thanks! Will merge once the build passes

**Comment by oleg-nenashev** [comment]
I definitely forgot about it :(

**Comment by oleg-nenashev** [comment]
Thanks @res0nance !

**Comment by daniel-beck** [comment]
How does https://github.com/jenkinsci/jenkins/blob/4404c6e662befe16e8daa1ea7a412df73b370086/core/src/main/java/jenkins/util/WindowsUtil.java#L25 even compile?

**Comment by res0nance** [comment]
@daniel-beck Wow now that you mention it I'm kinda amazed this works. Will refactor.

**Comment by jvz** [comment]
Ha, it works because Java only _recommends_ that the directory structure match the package name. That's not necessary, though it will likely break any code that assumes a one-to-one correspondence between a fully qualified class name and its respective filepath.

**Comment by daniel-beck** [comment]
@jvz So… we actually made this available as `hudson.os.WindowsUtil` to plugins…?

**Comment by jvz** [comment]
It appears so.

**Pull request merged**